File: lever/src/main/java/com/pin/lever/TargetAction.java
package com.pin.lever;

import android.graphics.Bitmap;
import android.graphics.drawable.Drawable;

final class TargetAction extends Action<Target> {

    TargetAction(Lever lever, Target target, Request data, int memoryPolicy, int networkPolicy,
                 Drawable errorDrawable, String key, Object tag, int errorResId) {
        super(lever, target, data, memoryPolicy, networkPolicy, errorResId, errorDrawable, key, tag,
                false);
    }

    @Override void complete(Bitmap result, Lever.LoadedFrom from) {
        if (result == null) {
            throw new AssertionError(
                    String.format("Attempted to complete action with no result!\n%s", this));
        }
        Target target = getTarget();
        if (target != null) {
            target.onBitmapLoaded(result, from);
            if (result.isRecycled()) {
                throw new IllegalStateException("Target callback must not recycle bitmap!");
            }
        }
    }

    @Override void error(Exception e) {
        Target target = getTarget();
        if (target != null) {
            if (errorResId != 0) {
                target.onBitmapFailed(e, lever.context.getResources().getDrawable(errorResId));
            } else {
                target.onBitmapFailed(e, errorDrawable);
            }
        }
    }
}

File: lever/src/main/java/com/pin/lever/TargetAction.kt
package com.pin.lever

import android.graphics.Bitmap
import android.graphics.drawable.Drawable
import com.pin.lever.Lever.LoadedFrom

internal class TargetAction(
    lever: Lever?,
    target: Target?,
    data: Request?,
    memoryPolicy: Int,
    networkPolicy: Int,
    errorDrawable: Drawable?,
    key: String?,
    tag: Any?,
    errorResId: Int
) : Action<Target?>(
    lever, target, data, memoryPolicy, networkPolicy, errorResId, errorDrawable, key, tag,
    false
) {
    public override fun complete(result: Bitmap?, from: LoadedFrom?) {
        if (result == null) {
            throw AssertionError(
                String.format(
                    "Attempted to complete action with no result!\n%s",
                    this
                )
            )
        }
        val target = getTarget()
        if (target != null) {
            target.onBitmapLoaded(result, from!!)
            check(!result.isRecycled) { "Target callback must not recycle bitmap!" }
        }
    }

    public override fun error(e: Exception?) {
        val target = getTarget()
        if (target != null) {
            if (errorResId != 0) {
                target.onBitmapFailed(e!!, lever!!.context.resources.getDrawable(errorResId))
            } else {
                target.onBitmapFailed(e!!, errorDrawable)
            }
        }
    }
}